File: BTreeCpp/BTreeCpp/BTree.cpp
#include "BTree.hpp"

#include <iostream>

Node::Node(int d, bool l) {
	nodeDegree = d;
	leaf = l;

	keys = new int[2 * d - 1];
	children = new Node * [2 * d];

	keysNum = 0;
}


void Node::__displayTree() {
	int i;
	for (i = 0; i < keysNum; i++) {
		if (!leaf) children[i]->__displayTree();
		std::cout << "->" << keys[i];
	}

	if (!leaf) children[i]->__displayTree();
}

void Node::__insertNonFull(int value) {
	int i = keysNum - 1;
	if (leaf) {
		while (i >= 0 && keys[i] > value) {
			keys[i + 1] = keys[i];
			i--;
		}

		keys[i + 1] = value;
		keysNum++;
	} else {
		while (i >= 0 && keys[i] > value) i--;

		if (children[i + 1]->keysNum == 2 * nodeDegree - 1) {
			__splitChild(children[i + 1], i + 1);

			if (keys[i + 1] < value) i++;
		}

		children[i + 1]->__insertNonFull(value);
	}
}

void Node::__splitChild(Node* node, int i) {
	Node* tempNode = new Node(node->nodeDegree, node->leaf);
	tempNode->keysNum = nodeDegree - 1;

	for (int j = 0; j < nodeDegree - 1; j++) 
		tempNode->keys[j] = node->keys[j + nodeDegree];

	if (!tempNode->leaf) {
		for (int j = 0; j < nodeDegree; j++) 
			tempNode->children[j] = node->children[j + nodeDegree];
	}

	node->keysNum = nodeDegree - 1;
	for (int j = keysNum; j >= i + 1; j--)
		children[j + 1] = children[j];

	children[i + 1] = tempNode;

	for (int j = keysNum - 1; j >= i; j--)
		keys[j + 1] = keys[i];

	keys[i] = node->keys[nodeDegree - 1];
	keysNum++;
}


void BTree::insertInTree(int value) {
	if (root == nullptr) {
		root = new Node(treeDegree, true);
		root->keys[0] = value;
		root->keysNum = 1;
	} else {
		if (root->keysNum == 2 * treeDegree - 1) {
			Node* newNode = new Node(treeDegree, false);

			newNode->children[0] = root;

			newNode->__splitChild(root, 0);

			int i = 0;
			if (newNode->keys[0] < value) i++;
			newNode->children[i]->__insertNonFull(value);

			root = newNode;
		} else root->__insertNonFull(value);
	}
}

void BTree::deleteKey(int value) {
	if (!root) {
		std::cerr << "\nThe BTree is empty!\n";
		return;
	}

	root->__deleteKey(value);

	if (root->keysNum == 0) {
		Node* tmp = root;
		if (root->leaf) root = NULL;
		else root = root->children[0];

		delete tmp;
	}
	return;
}

Node* Node::__searcKey(int key) {
	if (key == keys[0]) return this;
	int i = 0;
	while (i < keysNum && key > keys[i]) i++;

	if (keys[i] == key) return this;

	if (leaf) return nullptr;

	return children[i]->__searcKey(key);
}

int Node::__searcKey(int key, bool intKey) {
	int index = 0;
	while (index < keysNum && keys[index] < key) index++;
	return index;
}

void Node::__deleteKey(int key) {
	int index = __searcKey(key, true);
	if (index < keysNum && keys[index] == key) {
		if (leaf) __deleteFromLeaf(index);
		else __deleteFromNonLeaf(index);
	}
	else {
		if (leaf) {
			std::cerr << "\nThe key " << key << " is not exist in this BTree!\n";
			return;
		}

		bool status = (index == keysNum ? true : false);

		if (children[index]->keysNum < nodeDegree) __fill(index);

		if (status && index > keysNum) children[index - 1]->__deleteKey(key);
		else children[index]->__deleteKey(key);
	}
	return;
}

void Node::__deleteFromLeaf(int index) {
	for (int i = index + 1; i < keysNum; ++i) keys[i - 1] = keys[i];
	keysNum--;
	return;
}

void Node::__deleteFromNonLeaf(int index) {
	int key = keys[index];

	if (children[index]->keysNum >= nodeDegree) {
		int pred = __getPredecessor(index);
		keys[index] = pred;
		children[index]->__deleteKey(pred);
	}
	else if (children[index + 1]->keysNum >= nodeDegree) {
		int succ = __getSuccessor(index);
		keys[index] = succ;
		children[index + 1]->__deleteKey(succ);
	}
	else {
		__mergeKeys(index);
		children[index]->__deleteKey(key);
	}
	return;
}

void Node::__fill(int index) {
	if (index != 0 && children[index - 1]->keysNum >= nodeDegree) __borrowFromPrev(index);
	else if (index != keysNum && children[index + 1]->keysNum >= nodeDegree) __borrowFromNext(index);
	else {
		if (index != keysNum) __mergeKeys(index);
		else __mergeKeys(index - 1);
	}
	return;
}

int Node::__getPredecessor(int index) {
	Node* cur = children[index];
	while (!cur->leaf)
		cur = cur->children[cur->keysNum];

	return cur->keys[cur->keysNum - 1];
}

int Node::__getSuccessor(int index) {
	Node* cur = children[index + 1];
	while (!cur->leaf)
		cur = cur->children[0];

	return cur->keys[0];
}

void Node::__mergeKeys(int index) {
	Node* child = children[index];
	Node* sibling = children[index + 1];

	child->keys[nodeDegree - 1] = keys[index];
	for (int i = 0; i < sibling->keysNum; ++i) child->keys[i + nodeDegree] = sibling->keys[i];

	if (!child->leaf) {
		for (int i = 0; i <= sibling->keysNum; ++i)
			child->children[i + nodeDegree] = sibling->children[i];
	}


	for (int i = index + 1; i < keysNum; ++i)
		keys[i - 1] = keys[i];

	for (int i = index + 2; i <= keysNum; ++i)
		children[i - 1] = children[i];

	child->keysNum += sibling->keysNum + 1;
	keysNum--;

	delete (sibling);
	return;
}

void Node::__borrowFromPrev(int index) {
	Node* child = children[index];
	Node* sibling = children[index - 1];

	for (int i = child->keysNum - 1; i >= 0; --i)
		child->keys[i + 1] = child->keys[i];

	if (!child->leaf) {
		for (int i = child->keysNum; i >= 0; --i)
			child->children[i + 1] = child->children[i];
	}

	child->keys[0] = keys[index - 1];

	if (!child->leaf)
		child->children[0] = sibling->children[sibling->keysNum];

	keys[index - 1] = sibling->keys[sibling->keysNum - 1];

	child->keysNum += 1;
	sibling->keysNum -= 1;

	return;
}

void Node::__borrowFromNext(int index) {
	Node* child = children[index];
	Node* sibling = children[index + 1];

	child->keys[child->keysNum] = keys[index];

	if (!child->keysNum)
		child->children[child->keysNum + 1] = sibling->children[0];
	
	keys[index] = sibling->keys[0];

	for (int i = 1; i < sibling->keysNum; ++i)
		sibling->keys[i - 1] = sibling->keys[i];

	if (!sibling->leaf) {
		for (int i = 1; i <= sibling->keysNum; ++i)
			sibling->children[i - 1] = sibling->children[i];
	}

	child->keysNum += 1;
	sibling->keysNum -= 1;

	return;
}

File: BTreePython/BTree.py
class Node:
    def __init__(self, leaf=False):
        self.leaf = leaf

        self.keys = []
        self.children = []

class BTree(Node):
    def __init__(self, degree: int):
        self.root = Node(True)
        self.tree_degree = degree

    def __insert_in_not_full(self, node: Node, value: int):
        i = len(node.keys) - 1
        if node.leaf:
            node.keys.append((None, None))
            while i >= 0 and value[0] < node.keys[i][0]:
                node.keys[i + 1] = node.keys[i]
                i -= 1
            node.keys[i + 1] = value
        else:
            while i >= 0 and value[0] < node.keys[i][0]:
                i -= 1
            i += 1
            if i < len(node.children) and len(node.children[i].keys) == (2 * self.tree_degree) - 1:
                self.__split_child(node, i)
                if value[0] > node.keys[i][0]:
                    i += 1
            self.__insert_in_not_full(node.children[i], value)



    def __split_child(self, node: Node, i: int): 
        degree = self.tree_degree
        child = node.children[i]
        temp = Node(child.leaf)
        child.children.insert(i + 1, temp)
        node.keys.insert(i, child.keys[degree - 1])
        temp.keys = child.keys[degree: (2 * degree) - 1]
        child.keys = child.keys[0: degree - 1]
        if not child.leaf:
            temp.children = child.children[degree: 2 * degree]
            child.children = child.children[0: degree - 1]


    def delete_key(self, node: Node, value):
        t = self.tree_degree
        i = 0
        while i < len(node.keys) and value[0] > node.keys[i][0]:
            i += 1
        if node.leaf:
            if i < len(node.keys) and node.keys[i][0] == value[0]:
                node.keys.pop(i)
                return
            return

        if i < len(node.keys) and node.keys[i][0] == value[0]:
            return self.__delete_internal_node(node, value, i)
        elif len(node.child[i].keys) >= t:
            self.delete(node.child[i], value)
        else:
            if i != 0 and i + 2 < len(node.child):
                if len(node.child[i - 1].keys) >= t:
                    self.__delete_sibling(node, i, i - 1)
                elif len(node.child[i + 1].keys) >= t:
                    self.__delete_sibling(node, i, i + 1)
                else:
                    self.__delete_merge(node, i, i + 1)
            elif i == 0:
                if len(node.child[i + 1].keys) >= t:
                    self.__delete_sibling(node, i, i + 1)
                else:
                    self.__delete_merge(node, i, i + 1)
            elif i + 1 == len(node.child):
                if len(node.child[i - 1].keys) >= t:
                    self.__delete_sibling(node, i, i - 1)
                else:
                    self.__delete_merge(node, i, i - 1)
            self.delete(node.child[i], value)

    def __delete_internal_node(self, node: Node, value, i):
        t = self.tree_degree
        if node.leaf:
            if node.valueeys[i][0] == value[0]:
                node.keys.pop(i)
                return
            return

        if len(node.child[i].keys) >= t:
            node.keys[i] = self.__delete_predecessor(node.child[i])
            return
        elif len(node.child[i + 1].keys) >= t:
            node.keys[i] = self.__delete_successor(node.child[i + 1])
            return
        else:
            self.__delete_merge(node, i, i + 1)
            self.__delete_internal_node(node.child[i], value, self.tree_degree - 1)

    def __delete_predecessor(self, node: Node):
        if node.leaf:
            return node.keys.pop()
        n = len(node.keys) - 1
        if len(node.children[n].keys) >= self.tree_degree:
            self.__delete_sibling(node, n + 1, n)
        else:
            self.__delete_merge(node, n, n + 1)
        return self.____delete_predecessor(node.children[n])

    def __delete_successor(self, node: Node):
        if node.leaf:
            return node.keys.pop(0) if node.keys else None
        if len(node.children[1].keys) >= self.tree_degree:
            self.__delete_sibling(node, 0, 1)
        else:
            self.__delete_merge(node, 0, 1)
        return self.__delete_successor(node.children[0])


    def __delete_merge(self, node: Node, i, j):
        cnode = node.child[i]

        if j > i:
            rsnode = node.child[j]
            cnode.keys.append(node.keys[i])
            for k in range(len(rsnode.keys)):
                cnode.keys.append(rsnode.keys[k])
                if len(rsnode.child) > 0:
                    cnode.child.append(rsnode.child[k])
            if len(rsnode.child) > 0:
                cnode.child.append(rsnode.child.pop())
            new = cnode
            node.keys.pop(i)
            node.child.pop(j)
        else:
            lsnode = node.child[j]
            lsnode.keys.append(node.keys[j])
            for i in range(len(cnode.keys)):
                lsnode.keys.append(cnode.keys[i])
                if len(lsnode.child) > 0:
                    lsnode.child.append(cnode.child[i])
            if len(lsnode.child) > 0:
                lsnode.child.append(cnode.child.pop())
            new = lsnode
            node.keys.pop(j)
            node.child.pop(i)

        if node == self.root and len(node.keys) == 0:
            self.root = new

    def __delete_sibling(self, node: Node, i, j):
        cnode = node.child[i]
        if i < j:
            rsnode = node.child[j]
            cnode.keys.append(node.keys[i])
            node.keys[i] = rsnode.keys[0]
            if len(rsnode.child) > 0:
                cnode.child.append(rsnode.child[0])
                rsnode.child.pop(0)
            rsnode.keys.pop(0)
        else:
            lsnode = node.child[j]
            cnode.keys.insert(0, node.keys[i - 1])
            node.keys[i - 1] = lsnode.keys.pop()
            if len(lsnode.child) > 0:
                cnode.child.insert(0, lsnode.child.pop())
    
    def display_tree(self, node: Node, level = 0): 
        print(f"Level {level}: {len(node.keys)} keys: ", end='')
        for key in node.keys:
            print(key, end='')
        print()
        level += 1
        if len(node.children) > 0:
            for child in node.children:
                self.display_tree(child, level)

    def insert_in_tree(self, value):
        root = self.root
        if len(root.keys) == (2 * self.tree_degree) - 1:
            temp = Node()
            self.root = temp
            temp.children.insert(0, root)
            self.__split_child(temp, 0)
            self.__insert_in_not_full(temp, value)
        else: self.__insert_in_not_full(root, value)